packages/cozy-stack-client/src/OAuthClient.js

```diff
@@ -0,0 +1,358 @@
+   * @returns {boolean}
+   * @private
+   */
+  isRegistered() {
+    return this.oauthOptions.clientID !== ''
+  }
+
+  /**
+   * Converts a camel-cased data set to snake case, suitable for sending to the OAuth server
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  snakeCaseOAuthData(data) {
+    const mappedFields = {
+      softwareID: 'software_id',
+      softwareVersion: 'software_version',
+      clientID: 'client_id',
+      clientName: 'client_name',
+      clientKind: 'client_kind',
+      clientURI: 'client_uri',
+      logoURI: 'logo_uri',
+      policyURI: 'policy_uri',
+      notificationPlatform: 'notification_platform',
+      notificationDeviceToken: 'notification_device_token',
+      redirectURI: 'redirect_uris'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    // special case: turn redirect_uris into an array
+    if (
+      result['redirect_uris'] &&
+      result['redirect_uris'] instanceof Array === false
+    )
+      result['redirect_uris'] = [result['redirect_uris']]
+
+    return result
+  }
+
+  /**
+   * Converts a snake-cased data set to camel case, suitable for internal use
+   * @param   {object} data Initial data
+   * @returns {object} Formatted data
+   * @private
+   */
+  camelCaseOAuthData(data) {
+    const mappedFields = {
+      client_id: 'clientID',
+      client_name: 'clientName',
+      client_secret: 'clientSecret',
+      registration_access_token: 'registrationAccessToken',
+      software_id: 'softwareID'
+    }
+
+    const result = {}
+
+    Object.keys(data).forEach(fieldName => {
+      const key = mappedFields[fieldName] || fieldName
+      const value = data[fieldName]
+      result[key] = value
+    })
+
+    return result
+  }
+
+  /**
+   * Registers the currenly configured client with the OAuth server. 
+   * @throws {Error} When the client is already registered
+   * @returns {promise} A promise that resolves with a complete list of client information, including client ID and client secret.
+   */
+  async register() {
+    if (this.isRegistered()) throw new Error('Client already registered')
+
+    const data = await this.fetch(
+      'POST',
+      '/auth/register',
+      this.snakeCaseOAuthData({
+        redirectURI: this.oauthOptions.redirectURI,
+        clientName: this.oauthOptions.clientName,
+        softwareID: this.oauthOptions.softwareID,
+        clientKind: this.oauthOptions.clientKind,
+        clientURI: this.oauthOptions.clientURI,
+        logoURI: this.oauthOptions.logoURI,
+        policyURI: this.oauthOptions.policyURI,
+        softwareVersion: this.oauthOptions.softwareVersion,
+        notificationPlatform: this.oauthOptions.notificationPlatform,
+        notificationDeviceToken: this.oauthOptions.notificationDeviceToken
+      })
+    )
```

Comment:
I think that this is too much complicated. What if a property is missing? Can't you loop over every object's properties? I don't trust hard-coded exhaustive list.

Reply:
We do want an [exhaustive list though](https://github.com/cozy/cozy-stack/blob/master/docs/auth.md#post-authregister). We *could* just send the whole object because the server probably ignores the ones it doesn't know, but I don't think that's better.
And if a property is missing, it's totally ok.

Reply:
I mean, if the protocol changes and needs a new property, then the developer will add that property to their call, and the library won't take that new property into account without a change in the library code.